### ISSUE TITLE
`No query string was present` error on all requests using PersistedQuery since 2.0.18

### ISSUE THREAD

**Issue by ThomasF34** [rmosolgo/graphql-ruby#4389]
**Describe the bug**

Following the 2.0.18 update,  I have the error `"No query string was present"` on all requests.
(Note: we are using PersistedQuery)

**Versions**

- `graphql` version: `2.0.18` and `2.0.19`
- `rails`:  `7.0.4.3`
- other applicable versions:
  - `graphql-pro`: `1.24.10`
  - `graphql-batch`: `0.5.2`

**GraphQL query**

Here is an example of what happens

Request
```json
{
        "operationId":"sweep-webapp/3bf175adebc8f3cdedef06d22a9152ee",
        "operationName":"GetCurrentUser",
        "variables": {}
}
```

Response
```json
{
	"errors": [
		{
			"message": "No query string was present"
		}
	]
}
```

**Additional context**

- We are using PersistedQuery
- The operation with alias `3bf175adebc8f3cdedef06d22a9152ee` exists in DB
- The operation `last_used` timestamp is correctly updated 
- No change was done on our schema, only the update from 2.0.17 to 2.0.18 (or 2.0.19) introduced this bug 

Also, I found this PR (#4344) that [changed the parsing](https://github.com/rmosolgo/graphql-ruby/pull/4344/files#diff-32e3c2fdba5c368945b54604800adcca75ca505ab650327ffb6cc0611aeb79b0R378) of the query string, [used to trigger or not the error I encountered ](https://github.com/rmosolgo/graphql-ruby/pull/4344/files#diff-32e3c2fdba5c368945b54604800adcca75ca505ab650327ffb6cc0611aeb79b0R400)

Let me know if you need more info

**Comment by rmosolgo** [comment]
Yikes! Sorry for the trouble and thanks for the report. I'll take a look and follow up here soon.

**Comment by rmosolgo** [comment]
Could you share the full backtrace of the error, please? I expect there's some problem in ordering the different tracers, but I can't quite replicate it. The backtrace would include clues of what was called when the error was raised.

**Comment by ThomasF34** [comment]
Thanks for having a look

Unfortunatelly I had no stack trace as no error is raised. The error is only appearing in the request `errors` field and the only action seen in the backend is the update of the operation's `last_used_at` attribute

```ruby
Started POST "/api/q" for 172.18.0.1 at 2023-03-20 14:48:48 +0100
Processing by GraphqlController#execute as */*
  Parameters: {"operationId"=>"sweep-webapp/3bf175adebc8f3cdedef06d22a9152ee", "operationName"=>"GetCurrentUser", "variables"=>{}, "graphql"=>{"operationId"=>"sweep-webapp/3bf175adebc8f3cdedef06d22a9152ee", "operationName"=>"GetCurrentUser", "variables"=>{}}}
  User Load (1.5ms)  SELECT "users".* FROM "users" WHERE "users"."id" = $1 ORDER BY "users"."id" ASC LIMIT $2  [["id", 4], ["LIMIT", 1]]
  ↳ app/controllers/graphql_controller.rb:7:in `execute'
  Company Load (1.7ms)  SELECT "companies".* FROM "companies" WHERE "companies"."id" = $1 LIMIT $2  [["id", 1], ["LIMIT", 1]]
  ↳ config/initializers/warden.rb:2:in `block in <main>'
  GraphQL::Pro::OperationStore::ActiveRecordBackend::GraphQLClientOperation Load (6.3ms)  SELECT
              graphql_client_operations.id as id,
              graphql_client_operations.last_used_at as last_used_at,
              graphql_operations.id as operation_id,
              graphql_operations.name as operation_name,
              graphql_operations.body as operation_body,
              graphql_operations.digest as operation_digest
             FROM "graphql_client_operations" LEFT JOIN graphql_clients ON graphql_clients.id = graphql_client_operations.graphql_client_id
              LEFT JOIN graphql_operations ON graphql_operations.id = graphql_client_operations.graphql_operation_id WHERE "graphql_clients"."name" = $1 AND "graphql_client_operations"."alias" = $2 AND "graphql_client_operations"."is_archived" = $3 ORDER BY "graphql_client_operations"."id" ASC LIMIT $4  [["name", "sweep-webapp"], ["alias", "3bf175adebc8f3cdedef06d22a9152ee"], ["is_archived", false], ["LIMIT", 1]]
  ↳ app/controllers/graphql_controller.rb:27:in `execute'
Completed 200 OK in 28ms (Views: 0.1ms | ActiveRecord: 12.0ms | Allocations: 3069)


  GraphQL::Pro::OperationStore::ActiveRecordBackend::GraphQLClientOperation Update All (3.4ms)  UPDATE "graphql_client_operations" SET "last_used_at" = $1 WHERE "graphql_client_operations"."id" = $2  [["last_used_at", "2023-03-20 13:48:48.419376"], ["id", 482]]
 ```

**Comment by rmosolgo** [comment]
Oh -- my next guess is that there is some part of the tracing stack that is returning `nil` from `def parse(query_string:)`, which would cause `@document = ...` to be `nil` from the code you linked above. Could you please share the result of the following when using 2.0.18: 

```ruby
MyAppSchema.tracers 
MyAppSchema.trace_class.ancestors
```

That will tell me what tracers are active and then we can investigate their implementations.

**Comment by rmosolgo** [comment]
Ok, thanks for sharing that! (And derp -- sorry I had the wrong version for `.trace_class`.) I just released `graphql-pro` v1.24.1 with a fix for this issue. It requires `use GraphQL::Pro::OperationStore` to come _after_ `use GraphQL::Tracing::DataDogTrace`. Could you download that new version, update your schema definition, and let me know if it works properly for you? 

It turned out to be an oversight on my part for the cross-compatibility between tracing systems. Although `LegacyTrace` _does_ run legacy-style `def trace` methods, it always runs them _after_ new-style tracing modules. So, DataDogTracing was calling `query.selected_operation_name` which triggered parsing the query, but the OperationStore's `def trace` method hadn't run yet. Immediately after, `LegacyTrace` ran  OperationStore's `trace`, but it was too late, since parsing had already happened. 

To fix it, I added a new-style tracer module to OperationStore.  It will handle _this_ case, but if there are more issues like this, I'll consider a more robust approach to tracer ordering.

**Comment by suaron** [comment]
Hello @rmosolgo 

I have similar issue "No query string was present" with `graphql 2.0.21` and `graphql-pro 1.24.4` if `use GraphQL::Pro::OperationStore` placed before `use GraphQL::Tracing::PrometheusTracing` in scheme file.

Workaround is place `use GraphQL::Pro::OperationStore` after `use GraphQL::Tracing::PrometheusTracing`
